extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -126,19 +126,20 @@ public KafkaSupervisor(
 
 
   @Override
-  protected RecordSupplier<Integer, Long, KafkaRecordEntity> setupRecordSupplier()
+  protected RecordSupplier<KafkaTopicPartition, Long, KafkaRecordEntity> setupRecordSupplier()
   {
     return new KafkaRecordSupplier(
         spec.getIoConfig().getConsumerProperties(),
         sortingMapper,
-        spec.getIoConfig().getConfigOverrides()
+        spec.getIoConfig().getConfigOverrides(),
+        spec.getIoConfig().isMultiTopic()
     );
   }
 
   @Override
-  protected int getTaskGroupIdForPartition(Integer partitionId)
+  protected int getTaskGroupIdForPartition(KafkaTopicPartition partitionId)
   {
-    return partitionId % spec.getIoConfig().getTaskCount();
+    return partitionId.partition() % spec.getIoConfig().getTaskCount();
```

Comment:
Hmm, this seems a little tricky. 

This logic would always lead to a skew even with single topic ingestion, but the difference between most assigned taskGroupId and least assigned taskGroupId would have been only 1. With multiple topics and no limit on the number of topics, this difference can be anything.

Say there are 5 topics A, B, C, D, E with 6 partitions each and task count is 4.
Then taskGroupId 0 and 1 would get 10 partitions each
whereas taskGroupId 2 and 3 would get 5 partitions each.

The skew can be reduced in two ways:
- Get total partition count of each topic and then assign them new integer ids. e.g. A:1 -> 1, A:2 -> 2... A:5 -> 5, then B:1 -> 6, B:2 -> 7 and so on. Then do a module of these new overall IDs instead of the `partitionId.partition()`. This is very cumbersome though.
- A simpler approach could be to offset the assignment for every topic. So A starts assignment at taskGroupId 0, but B starts assignment at taskGroupId 1 and so on.
  - While constructing this class, we just need to get the list of topics from the stream.
  - Then the code here could become
```
topics = List of topics
return (partitionId.partition() + topics.getIndexOf(partitionId.topic())) % taskCount;
```

Reply:
Wow. Nice catch. I am wary of that logic as I don't want the assignment to change if someone adds one more topic to the stream. I am thinking something like below 
`(31*topic.hashCode() + partition) % taskCount. ` but that may not lead to an optimal assignment.